### PR TITLE
script: regenesis

### PR DESCRIPTION
**Pull request by tynes**
**Description**
Adds the state surgery script required for regenesis. This depends on the contracts being build and served with `yarn serve`

```bash
$ STATE_DUMP_PATH=http://127.0.0.1:8081/state-dump.latest.json \
    SEQUENCER_ENDPOINT=http://127.0.0.1:8547 \
    node state-surgery.js  | jq . | tee state-dump.json
```

### CODE REVIEW

File: scripts/state-surgery.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * Required env vars:
+ *  SEQUENCER_ENDPOINT
+ *  STATE_DUMP_PATH
+ */
+
+const fs = require('fs');
+const axios = require('axios');
+const { JsonRpcProvider } = require('@ethersproject/providers');
+const { getContractDefinition } = require('@eth-optimism/contracts')
+const { keccak256 } = require("@ethersproject/keccak256");
```
Comment (tynes) [inline scripts/state-surgery.js]: this is dead code now

File: scripts/state-surgery.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        nonce: account.nonce,
+        code: contracts.ProxyEOA.deployedBytecode,
+        storage: account.storage,
+        abi: contracts.ProxyEOA.abi
+      }
+    } else if (isPredeploy(address) || isSystemAccount(address)) {
+      // Predeploys and System Accounts keep the same nonce and code
+      // Iterate through the contractsDump to find the matching addresses so
+      // that the storage can be pulled in
+      for (const [name, dumpAccount] of Object.entries(contractsDump.accounts)) {
```
Comment (annieke) [inline scripts/state-surgery.js]: are we worried about the performance of this double loop, or is there not a lot of these?

Comment (tynes) [reply scripts/state-surgery.js]: This ended up not being required

File: scripts/state-surgery.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        storage: account.storage,
+        abi: []
+      }
+    }
+  }
+  contractsDump.unknowns = unknowns
+  console.log(JSON.stringify(contractsDump))
+})().catch(err => {
+  console.error(err)
+  console.log(JSON.stringify({}))
```
Comment (snario) [inline scripts/state-surgery.js]: ?

File: scripts/state-surgery.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  return int <= 9 && int !== 0
+}
+
+function getFindAndReplacedCode(str) {
+  return str.split(
+    '336000905af158601d01573d60011458600c01573d6000803e3d621234565260ea61109c52'
+  ).join(
+    '336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b'
+  )
+}
```
Comment (snario) [inline scripts/state-surgery.js:142]: 🤔 @ben-chain

Comment (tynes) [reply scripts/state-surgery.js:142]: https://github.com/ethereum-optimism/solidity/issues/21

**Pull request merged**